ThirdPartyAdapters/applovin/applovin/src/main/java/com/applovin/mediation/AppLovinNativeAdListener.java

```diff
@@ -0,0 +1,107 @@
+                                    AppLovinSdk sdk,
+                                    Context context) {
+        this.adapter = adapter;
+        this.nativeListener = nativeListener;
+        this.appLovinSdk = sdk;
+        contextWeakReference = new WeakReference<>(context);
+    }
+
+    @Override
+    public void onNativeAdsLoaded(List<AppLovinNativeAd> nativeAds) {
+        if (nativeAds.size() > 0 && isValidNativeAd(nativeAds.get(0))) {
+            appLovinSdk.getNativeAdService()
+                    .precacheResources(nativeAds.get(0), AppLovinNativeAdListener.this);
+        } else {
+            notifyAdFailure(AdRequest.ERROR_CODE_NO_FILL);
+        }
+    }
+
+    @Override
+    public void onNativeAdsFailedToLoad(final int errorCode) {
+        notifyAdFailure(AppLovinUtils.toAdMobErrorCode(errorCode));
+    }
+
+    @Override
+    public void onNativeAdImagesPrecached(AppLovinNativeAd ad) {
+        // Create a native ad.
+        if (contextWeakReference.get() == null) {
+            Log.w(AppLovinNativeAdapter.TAG, "Failed to create mapper. Context is null.");
+            notifyAdFailure(AdRequest.ERROR_CODE_INTERNAL_ERROR);
+            return;
+        }
+        final AppLovinNativeAdMapper mapper =
+                new AppLovinNativeAdMapper(ad, contextWeakReference.get());
+        AppLovinSdkUtils.runOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+                nativeListener.onAdLoaded(adapter, mapper);
+            }
+        });
+    }
+
+    @Override
+    public void onNativeAdVideoPreceached(AppLovinNativeAd ad) {
+        // Do nothing.
+    }
+
+    @Override
+    public void onNativeAdImagePrecachingFailed(AppLovinNativeAd ad, final int errorCode) {
+        notifyAdFailure(AppLovinUtils.toAdMobErrorCode(errorCode));
+    }
+
+    @Override
+    public void onNativeAdVideoPrecachingFailed(AppLovinNativeAd ad, final int errorCode) {
+        notifyAdFailure(AppLovinUtils.toAdMobErrorCode(errorCode));
+    }
+
+    /**
+     * Sends a failure callback to {@link #nativeListener}.
+     * @param errorCode AdMob {@link AdRequest} error code.
+     */
+    private void notifyAdFailure(final int errorCode) {
+        AppLovinSdkUtils.runOnUiThread(new Runnable() {
+            @Override
+            public void run() {
+                nativeListener.onAdFailedToLoad(adapter, errorCode);
+            }
+        });
+
+    }
+
+    /**
+     * Checks whether or not the {@link AppLovinNativeAd} has all the required assets.
+     * @param nativeAd AppLovin native ad.
+     * @return {@code true} if the native ad has all the required assets.
+     */
+    private static boolean isValidNativeAd(AppLovinNativeAd nativeAd) {
+        return nativeAd.isImagePrecached() && nativeAd.getImageUrl() != null
+                && nativeAd.getIconUrl() != null && nativeAd.getTitle() != null
+                && nativeAd.getDescriptionText() != null && nativeAd.getCtaText() != null;
```

Comment:
In regards to `isImagePrecached()` - It's not always true that AppLovin SDK precaches the image. Check out setting `NATIVE_ADS_IN_BATCH_TO_PRECACHE`. in `SettingTypes.java`.

Reply:
I'm explicitly requesting the ads to be pre-cached once and ad is loaded.